astropy/table/table.py

```diff
@@ -2021,6 +2021,70 @@ def group_by(self, keys):
 
         return groups.table_group_by(self, keys)
 
+    def to_pandas(self):
+        """
+        Return a :class:`pandas.DataFrame` instance
+
+        Returns
+        -------
+        dataframe : :class:`pandas.DataFrame`
+            A pandas :class:`pandas.DataFrame` instance
+        """
+        from pandas import DataFrame
+
+        if self.has_mixin_columns:
+            raise ValueError("Cannot convert a table with mixin columns to a pandas DataFrame")
+
+        for column in self.columns:
+            if len(self.dtype[column].shape) > 0:
+                raise ValueError("Cannot convert a table with multi-dimensional columns to a pandas DataFrame")
+
+        if self.masked:
+            # We have to add the columns manually so as to treat the mask
+            # correctly
+            d = DataFrame()
+            for name, column in self.columns.items():
+                if isinstance(column, MaskedColumn):
+                    if column.dtype.kind in ['i']:
+                        d[name] = column.astype(float).filled(np.nan)
+                    elif column.dtype.kind in ['f', 'c']:
+                        d[name] = column.filled(np.nan)
+                    else:
+                        d[name] = column.astype(np.object).filled(np.nan)
```

Comment:
Missing an `else:` clause here if not `isinstance(column, MaskedColumn)`.  (Though without mixins that can't actually happen, but good to future-proof things).  Inspired by @mdboom, I've been using an idiom of collecting columns in a dict prior to making the table.  So something like this might generically work and doesn't require an intermediate numpy array in the non-masked case.  This also could support special-casing for certain mixins like you mentioned, e.g. Time => datetime:
```
out = OrderedDict()
for name, column in self.columns.items():
    if isinstance(column, MaskedColumn):
        ....
    elif isinstance(column, Time):
        out[name] = column.datetime  # this actually works in pandas, don't know if there is a more efficient way though
    else:  # Not a masked column
        out[name] = column

return DataFrame(out)
```